## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [190]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [191]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [192]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [193]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.shape
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [194]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [195]:
# For the category with the most entries, which subcategory had the most entries?
df[df['category'] == 'consumables']['subcategory'].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [196]:
# Which five clients had the most entries in the data?
df['client_id'].value_counts().head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [197]:
# Store the client ids of those top 5 clients in a list.
clients_top = df['client_id'].value_counts().head(5).index.tolist()

In [198]:
# How many total units (the qty column) did the client with the most entries order order?
df[df['client_id'] == clients_top[0]]['qty'].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [199]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [200]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def shipping_price(row):
    weight_qty = row['unit_weight'] * row['qty']
    if row['unit_weight'] > 50:
        return 7.00 * weight_qty
    else:
        return 10.00 * weight_qty
    
df['shipping_price'] = df.apply(shipping_price, axis=1)


In [201]:
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  \
0     1096.80     762.71         7.50  105            1  115164.00   
1       24.95      15.09         1.49   21            0     523.95   
2       13.52       7.86         1.68   39            6     527.28   
3       36.42      24.85         1.23   29            3    1056.18   
4      195.10     108.17        46.43   20            1    3902.00   

   shipping_price  
0          7875.0  
1           312.9  
2           655.2  
3           356.7  
4          9286.0

In [202]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
total_with_ship = df['subtotal'] + df['shipping_price']
df['total_price'] = total_with_ship + (total_with_ship * 0.0925)

In [203]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['cost_of_each_line'] = (df['unit_cost'] * df['qty']) + df['shipping_price']

In [204]:
# Create a column for the profit of each line using line cost and line price
df['profit'] = df['total_price'] - df['cost_of_each_line']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  shipping_price    total_price  cost_of_each_line  \
0            1  115164.00          7875.0  134420.107500           87959.55   
1            0     523.95           312.9     914.258625             629.79   
2            6     527.28           655.2    1291.859400             961.74   
3            3    1056.18           356.7    1543.571400            1077.35   
4            1    3902.00          9286.0   14407.890000           11449.40   

         profit  
0  46460.557500  
1    284.468625  
2    330.119400  
3    466.221400  
4   2958.490000  

[5 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [205]:
# Check your work using the totals above
orders_1 = df.loc[df['order_id'] == 2742071]['total_price'].sum()
print(orders_1)

orders_2 = df.loc[df['order_id'] == 2173913]['total_price'].sum()
print(orders_2)

orders_3 = df.loc[df['order_id'] == 6128929]['total_price'].sum()
print(orders_3)

158105.649525
180488.05755
996593.762225


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [206]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
for client in clients_top:
    total_spent = df.loc[df['client_id'] == client]['total_price'].sum()
    print(f'Client {client} spent ${total_spent}')


Client 33615 spent $9228994.707925
Client 66037 spent $10908841.7939
Client 46820 spent $10489716.350274999
Client 38378 spent $14508053.404375
Client 24741 spent $84663989.704375


In [207]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
df_summary = pd.DataFrame(columns=['client_id', 'qty', 'total_shipping_price', 'total_line_price', 'total_revenue', 'total_profit'])
df_summary['client_id'] = clients_top
df_summary['qty'] = df.groupby('client_id')['qty'].sum().loc[clients_top].values
df_summary['total_shipping_price'] = df.groupby('client_id')['shipping_price'].sum().loc[clients_top].values
df_summary['total_line_price'] = df.groupby('client_id')['cost_of_each_line'].sum().loc[clients_top].values
df_summary['total_revenue'] = df.groupby('client_id')['total_price'].sum().loc[clients_top].values
df_summary['total_profit'] = df.groupby('client_id')['profit'].sum().loc[clients_top].values
df_summary


client_id     qty  total_shipping_price  total_line_price  total_revenue  \
0      33615   64313             2608560.3        6954889.32   9.228995e+06   
1      66037   43018             1989501.5        7598832.63   1.090884e+07   
2      46820   75768             2284215.0        7689957.80   1.048972e+07   
3      38378   73667             4895361.6       11100627.18   1.450805e+07   
4      24741  239862             7318757.5       47881208.84   8.466399e+07   

   total_profit  
0  2.274105e+06  
1  3.310009e+06  
2  2.799759e+06  
3  3.407426e+06  
4  3.678278e+07

In [208]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 
money_columns = ['Client ID', 'Units', 'Shipping (millions)', 'Total Revenue (millions)', 'Total Cost (millions)', 'Total Profit (millions)']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(x):
    if type(x) == int or type(x) == float:
        return x / 1000000
    else:
        print(f'Error: {x} is not a number.')
        return x



# Apply the currency_format_millions function to only the money columns.
df_summary['total_shipping_price'] = df_summary['total_shipping_price'].map(currency_format_millions)
df_summary['total_line_price'] = df_summary['total_line_price'].map(currency_format_millions)
df_summary['total_revenue'] = df_summary['total_revenue'].map(currency_format_millions)
df_summary['total_profit'] = df_summary['total_profit'].map(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
df_summary.columns = money_columns
df_summary



Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      33615   64313             2.608560                  6.954889   
1      66037   43018             1.989502                  7.598833   
2      46820   75768             2.284215                  7.689958   
3      38378   73667             4.895362                 11.100627   
4      24741  239862             7.318758                 47.881209   

   Total Cost (millions)  Total Profit (millions)  
0               9.228995                 2.274105  
1              10.908842                 3.310009  
2              10.489716                 2.799759  
3              14.508053                 3.407426  
4              84.663990                36.782781

In [210]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
df_summary.sort_values('Total Profit (millions)', ascending=False, inplace=True)
df_summary

Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
4      24741  239862             7.318758                 47.881209   
3      38378   73667             4.895362                 11.100627   
1      66037   43018             1.989502                  7.598833   
2      46820   75768             2.284215                  7.689958   
0      33615   64313             2.608560                  6.954889   

   Total Cost (millions)  Total Profit (millions)  
4              84.663990                36.782781  
3              14.508053                 3.407426  
1              10.908842                 3.310009  
2              10.489716                 2.799759  
0               9.228995                 2.274105

In [ ]:
# Write a brief 2-3 sentence summary of your findings.
# The top 5 clients spent almost 10 million each. The total profit was close to 50 million.